File: 0349_intersection_of_two_arrays.py
from typing import List
# n = len(num1)
# m = len(num2)


class Solution:
    # Time: O(max(n*2, m*2))
    # Space: O(n + m)
    def intersection(self, nums1: List[int], nums2: List[int]) -> List[int]:
        dict = {}
        intersection = []
        for n1 in nums1:
            if n1 not in dict.keys():
                dict[n1] = 1
        for n2 in nums2:
            if n2 in dict.keys() and n2 not in intersection:
                intersection.append(n2)
        return intersection

class Solution2:
    # Time: O(n * m)
    # Space: O(min(n, m))
    def intersection(self, nums1: List[int], nums2: List[int]) -> List[int]:
        common_elements = set()
        for num in nums1:
            if num in nums2:
                common_elements.add(num)
        return list(common_elements)

class Solution3:
    # Time: O(n + m)
    # Space: O(n + m)
    def intersection(self, nums1: List[int], nums2: List[int]) -> List[int]:
        set1 = set(nums1)
        set2 = set(nums2)
        return list(set1 & set2)

File: go/0349_intersection_of_two_arrays.go
package main

// n = len(nums1), m = len(nums2)
// Time:  O(n + m)
// Space: O(n)
func intersection(nums1 []int, nums2 []int) []int {
	intersections := []int{}
	seen := make(map[int]bool)

	for _, n := range nums1 {
		seen[n] = true
	}

	for _, n := range nums2 {
		if seen[n] {
			intersections = append(intersections, n)
			delete(seen, n)
		}
	}

	return intersections
}
